nitransforms/resampling.py

```diff
@@ -7,57 +7,126 @@
+
+from os import cpu_count
+from concurrent.futures import ProcessPoolExecutor, as_completed
 from pathlib import Path
+from typing import Tuple
+
 import numpy as np
 from nibabel.loadsave import load as _nbload
+from nibabel.arrayproxy import get_obj_dtype
+from nibabel.spatialimages import SpatialImage
 from scipy import ndimage as ndi
 
 from nitransforms.base import (
     ImageGrid,
+    TransformBase,
     TransformError,
     SpatialReference,
     _as_homogeneous,
 )
 
+SERIALIZE_VOLUME_WINDOW_WIDTH: int = 8
+"""Minimum number of volumes to automatically serialize 4D transforms."""
+
+
+def _apply_volume(
+    index: int,
+    data: np.ndarray,
+    targets: np.ndarray,
+    order: int = 3,
+    mode: str = "constant",
+    cval: float = 0.0,
+    prefilter: bool = True,
+) -> Tuple[int, np.ndarray]:
+    """
+    Decorate :obj:`~scipy.ndimage.map_coordinates` to return an order index for parallelization.
+
+    Parameters
+    ----------
+    index : :obj:`int`
+        The index of the volume to apply the interpolation to.
+    data : :obj:`~numpy.ndarray`
+        The input data array.
+    targets : :obj:`~numpy.ndarray`
+        The target coordinates for mapping.
+    order : :obj:`int`, optional
+        The order of the spline interpolation, default is 3.
+        The order has to be in the range 0-5.
+    mode : :obj:`str`, optional
+        Determines how the input image is extended when the resamplings overflows
+        a border. One of ``'constant'``, ``'reflect'``, ``'nearest'``, ``'mirror'``,
+        or ``'wrap'``. Default is ``'constant'``.
+    cval : :obj:`float`, optional
+        Constant value for ``mode='constant'``. Default is 0.0.
+    prefilter: :obj:`bool`, optional
+        Determines if the image's data array is prefiltered with
+        a spline filter before interpolation. The default is ``True``,
+        which will create a temporary *float64* array of filtered values
+        if *order > 1*. If setting this to ``False``, the output will be
+        slightly blurred if *order > 1*, unless the input is prefiltered,
+        i.e. it is the result of calling the spline filter on the original
+        input.
+
+    Returns
+    -------
+    (:obj:`int`, :obj:`~numpy.ndarray`)
+        The index and the array resulting from the interpolation.
+
+    """
+    return index, ndi.map_coordinates(
+        data,
+        targets,
+        order=order,
+        mode=mode,
+        cval=cval,
+        prefilter=prefilter,
+    )
+
 
 def apply(
-    transform,
-    spatialimage,
-    reference=None,
-    order=3,
-    mode="constant",
-    cval=0.0,
-    prefilter=True,
-    output_dtype=None,
-):
+    transform: TransformBase,
+    spatialimage: str | Path | SpatialImage,
+    reference: str | Path | SpatialImage = None,
+    order: int = 3,
+    mode: str = "constant",
+    cval: float = 0.0,
+    prefilter: bool = True,
+    output_dtype: np.dtype = None,
+    serialize_nvols: int = SERIALIZE_VOLUME_WINDOW_WIDTH,
```

Comment:
There's no parameter in the docstring, and the whole concept of serialization here is non-obvious. To the best of my understanding, the idea is that you want to parallelize over volumes but only if you hit a threshold that justifies the overhead of multiprocessing.

Have you compared the performance of multiprocessing versus async workers? If you're able to async or thread, you will have much less overhead and might not need this additional concept.

Reply:
Overall it is a great interpretation.

There's something not so explicit: if you want real 4D interpolation (e.g., head motion + slice timing corrections) then you want to set `serialize_nvols = np.inf` because you can't split by volume (we've talked about doing this in a moving window, and then this parameter would map onto the width of that window).

Happy to document better.

Reply:
I have added this parameter's description in the docstring, as it was missing. LMKWYT